### PR TITLE
[JENKINS-39232] Walk the DelegatingComputerLauncher instances when checking if JNLPComputerLauncher

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-39232](https://issues.jenkins-ci.org/browse/JENKINS-39232)

@jenkinsci/code-reviewers @reviewbybees

Replaces #2601

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
I do not believe it's a perfect fix since anybody can implement his own launcher using `ComputerLauncher` API. And there are many examples: https://github.com/search?q=org%3Ajenkinsci+%22extends+computerlauncher%22&type=Code

I agree it worths adding support of Delegating launchers, but it does not fix the issue

File: core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java [hunk 1]
```diff
@@ -37,6 +41,16 @@
 @Extension
 public class DefaultJnlpSlaveReceiver extends JnlpAgentReceiver {
 
+    /**
+     * Disables strict verification of connections. Turn this on if you have plugins that incorrectly extend
+     * {@link ComputerLauncher} when then should have extended {@link DelegatingComputerLauncher}
+     *
+     * @since FIXME
+     */
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java]: `@Extension(NoExternalUse.class)`

File: core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java [hunk 2]
```diff
@@ -37,20 +44,60 @@
...
         Computer computer = Jenkins.getInstance().getComputer(clientName);
         return computer != null;
     }
 
+    private static ComputerLauncher getDelegate(ComputerLauncher launcher) {
+        try {
+            Method getDelegate = launcher.getClass().getMethod("getDelegate");
+            return ComputerLauncher.class.isAssignableFrom(getDelegate.getReturnType()) ? (ComputerLauncher) getDelegate
+                    .invoke(launcher) : null;
+        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java]: ReflectiveOperationException?

File: core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java [hunk 3]
```diff
@@ -37,20 +44,60 @@
...
             return;
         }
+        ComputerLauncher launcher = computer.getLauncher();
+        while (!(launcher instanceof JNLPLauncher)) {
+            ComputerLauncher l;
+            if (launcher instanceof DelegatingComputerLauncher) {
+                launcher = ((DelegatingComputerLauncher) launcher).getLauncher();
+            } else if (launcher instanceof ComputerLauncherFilter) {
+                launcher = ((ComputerLauncherFilter) launcher).getCore();
+            } else if (null != (l = getDelegate(launcher))) {  // TODO remove when all plugins are fixed
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java]: Does not guarantee a success for old plugins from what I see

File: core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java [hunk 4]
```diff
@@ -37,20 +44,60 @@
...
+                launcher = l;
+            } else {
+                if (disableStrictVerification) {
+                    LOGGER.log(Level.WARNING, "Connecting {0} as a JNLP agent where the launcher does not mark itself "
+                            + "correctly as being a JNLP agent", clientName);
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java]: Please explicitly provide a class of the launcher

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
If we want to go in such way, we need better diagnostic messages. :bug: for it.

I still have concerns about such approach due to impact on plugins, but it's a subject for voting by @jenkinsci/core

**Review by oleg-nenashev — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
I approve the code change itself since it makes the situation better that it is in 2.27. So 🐝 and 👍 as a partial fix.

But I still do not feel it's a right approach to nuke so many instances. Hence as a community member I keep 👎 regarding the change as a **complete** fix of JENKINS-39232

**Comment by oleg-nenashev** [comment]
@reviewbybees done
@jenkinsci/code-reviewers We really really really want to get it in the next weekly

**Comment by oleg-nenashev** [comment]
P.S: I will follow-up on the default behavior

**Comment by oleg-nenashev** [comment]
I will merge it into the next weekly if nobody shouts

File: core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java [hunk 5]
```diff
@@ -37,20 +44,60 @@
...
+            SystemProperties.getBoolean(DefaultJnlpSlaveReceiver.class.getName() + ".disableStrictVerification");
+
+
     @Override
     public boolean owns(String clientName) {
         Computer computer = Jenkins.getInstance().getComputer(clientName);
         return computer != null;
     }
 
+    private static ComputerLauncher getDelegate(ComputerLauncher launcher) {
```
Comment (KostyaSha) [inline core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java]: comment wasn't addressed?

File: core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java [hunk 6]
```diff
@@ -37,20 +46,82 @@
...
+        try {
+            Method getDelegate = launcher.getClass().getMethod("getDelegate");
+            if (ComputerLauncher.class.isAssignableFrom(getDelegate.getReturnType())) {
+                return (ComputerLauncher) getDelegate.invoke(launcher);
+            }
+        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
+            // ignore
```
Comment (KostyaSha) [inline core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java]: you can name `ignore` instead `e` to disable warnings static analysing tools.

**Pull request merged**